server/storage: Add admin endpoint to manually run range through queue

a-robinson opened the pull request:
Fixes part of #25073

Release note: Added a new admin server endpoint `/_admin/v1/queue` which
runs the specified range through the specified internal queue, with the
optional `?skip_should_queue=true` parameter telling the system to
blindly run without first checking whether it needs to be run. Intended
primarily for debugging purposes.

a-robinson committed:
server/storage: Add admin endpoint to manually run range through queue

Fixes part of #25073

Release note (general change): Added a new admin server endpoint
`/_admin/v1/enqueue_range` which runs the specified range through the
specified internal queue on the specified node (or all nodes). The
skip_should_queue parameter can also be specified to tell the system to
blindly run without first checking whether it needs to be run. Intended
primarily for debugging purposes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26554)
<!-- Reviewable:end -->

tschottdorf commented on pkg/storage/store.go:
```diff
@@ -4233,6 +4234,70 @@ func (s *Store) AllocatorDryRun(
+	ctx context.Context, queueName string, repl *Replica, skipShouldQueue bool,
+) ([]tracing.RecordedSpan, string, error) {
+	var queue queueImpl
+	switch strings.ToLower(queueName) {
```
Not required, but could this loop through all the queues registered with the scanner and match on the name? Then we'd not have to change anything when adding queues/changing a queue name.

tschottdorf commented on pkg/storage/store.go:
```diff
@@ -4233,6 +4234,70 @@ func (s *Store) AllocatorDryRun(
+		return nil, "", pErr.GoError()
+	}
+	if !hasLease {
+		return nil, "", nil
```
No message here?

tbg reviewed (commented):
Looks good so far! I'll give a closer review when you've pushed the next set of updates.
Do you think it's also useful to process a replica through a given queue on a non-leaseholder?

tschottdorf commented:
@a-robinson we may also consider not doing the POST stuff here if we're going to do #26551 anyway. Will take me a bit to get that in, though we're likely not going to backport this change anyway, right?

a-robinson commented:
> @a-robinson we may also consider not doing the POST stuff here if we're going to do #26551 anyway.

I thought #26551 was just a way of taking status/admin server commands and making them accessible via SQL. Why would that change the decision of whether the admin server command should be a GET or a POST?

> Will take me a bit to get that in, though we're likely not going to backport this change anyway, right?

I haven't been planning on backporting it.

a-robinson commented:
> @a-robinson I was using the grpc endpoints, not the http ones, and so the problem would go away.

Not if we also want the endpoint to be usable via HTTP, which certainly seems like it'd be nice.

tschottdorf commented:
Yep, it would be nice, though I'm not sure there are many situations in which I'd personally use that over the SQL version. Either way, I'm not saying don't do it. Just, if it's a total pain and there's a string of unrelated issues, we could make do without it for now.

a-robinson commented:
I'd use that over the SQL version in all situations where we have UI access but not SQL access. I suppose we could use a local grpc client program to get the same effect, but it seems easier to just send a POST in such cases.

Anyway, I've update the PR to address all comments. TFTRs!

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/server/serverpb/admin.proto, line 758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEbVVgJ6RXomZvP89Nx:-LEooEjE6VMLWJJ0YuVj:b-8s30xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/server/serverpb/admin.proto#L758)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I generally prefer to just use query parameters instead of building operation-specific path patterns. 
</blockquote></details>

I thought the path pattern was more popular in REST API design, and there are a handful of other admin endpoints here that do something similar. I've changed it, though.

---

*[pkg/storage/store.go, line 4245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEXar4D_-9oEWUyaIDK-r1-4245:-LEkr8nb4zpDUTtG2V9T:b-afh902) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/storage/store.go#L4245)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Not required, but could this loop through all the queues registered with the scanner and match on the name? Then we'd not have to change anything when adding queues/changing a queue name.
</blockquote></details>

Done, although in order to make it less ridiculous I changed the name of the consistency queue from "replica consistency checker" to just "consistencyChecker". I'd go all the way down to "consistency", but do you think that would look confusing in the logs when it's used to annotate contexts?

---

*[pkg/storage/store.go, line 4270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEbVCV0BhbYmPbUNJPQ:-LEpQrKA7UxVygk2c5BI:b-wxlmwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/storage/store.go#L4270)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What about the queues that do run on non-leaseholders? Is there a way to override this? Maybe look at the queue itself to see if it requires the lease?
</blockquote></details>

Done. We'll now run on all nodes that have the replica if the lease isn't required.

---

*[pkg/storage/store.go, line 4277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEXar4D_-9oEWUyaIDK-r1-4277:-LEkrBKO-f120WgFSOzy:b-k1h98t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/storage/store.go#L4277)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

No message here?
</blockquote></details>

That was intentional in order to make sure the most relevant error shines through. If the leaseholder returns an error, we wouldn't want it to be covered up by an error from a non-leaseholder that just says it isn't the leaseholder. This is, of course, optimistically assuming that one of the replicas is able to get the lease. If that isn't true, something bigger is probably wrong.

But in order to allow running on replicas that aren't leaseholder for queues that don't require the lease, I've changed a few things around and am now returning a message here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LEonsev-DbIt6hcowzL:b-xtz20d)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
> Looks good! How about a simple unit test? 
 Done.

> And add something to the debug overview page that gives the relevant curl invocations for an example or two.

Done, but I think I've committed crimes against typescript in the process, so cc @couchand to review that part.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/server/admin.go, line 1593 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEbUfvg5YD8PzE7l1CR:-LEzav3j6uCyjZJnuJqK:bld540p) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/server/admin.go#L1593)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Maybe the request should contain a key instead of an ID, then. Or a specific node ID.
</blockquote></details>

Sorry for not responding to this yesterday. In my opinion that makes the interface much tougher to use. The only other status/admin server endpoint that takes keys is `SpanStats`, which I don't think is ever used directly by a human. When debugging, we (or at least I) usually work in terms of range IDs, at least partially due to the design of the existing debug endpoints. It would also add confusion about which types of keys are usable (how do they have to be encoded? pretty keys only?).

I can switch it if you prefer, but I think after the recent changes to support running on non-leaseholder queues that the problem isn't as bad as it was before.

---

*[pkg/server/serverpb/admin.proto, line 553 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEy-kSb54Qw-mukSEme:-LEzcHf62Bm3gcSh0b8F:b-j5yt41) ([raw file](https://github.com/cockroachdb/cockroach/blob/86f0469323c5a2cba0b08e1a6663a96d35250caf/pkg/server/serverpb/admin.proto#L553)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This is a bit more brute force than it needs to be (could also look up the range descriptor), but I like that this can work even if the meta ranges are wedged (and perhaps the queue is required to unwedge it). Consider putting that in a comment.
</blockquote></details>

As referenced the comment that Ben commented on in the original set of changes to admin.go, we can't look up the range descriptor from the range ID. We'd have to change the endpoint to take a key rather than a range ID in order to do a range descriptor lookup, which in my opinion makes the endpoint tougher to use. If I'm outvoted, though, I'm happy to change.

---

*[pkg/server/serverpb/admin.proto, line 557 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEy-v4-Dempxl3JuUhh:-LEzdDji13qNN4x0AIkp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/86f0469323c5a2cba0b08e1a6663a96d35250caf/pkg/server/serverpb/admin.proto#L557)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Should explain what this is matched against.
</blockquote></details>

Done.

---

*[pkg/server/serverpb/admin.proto, line 558 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEy-znNCRT3TK6_7BWq:-LEzcp3VBW43cZH4b4b5:byrhjpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/86f0469323c5a2cba0b08e1a6663a96d35250caf/pkg/server/serverpb/admin.proto#L558)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Isn't this TODO done?
</blockquote></details>

Yes, thanks. Although I can't say I understand why it works now but it didn't work in my original version of this change.

---

*[pkg/server/serverpb/admin.proto, line 766 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEy08psC_kTa45DX4c_:-LEzdQjPAOd6oZqjwu_o:b-jbu5lu) ([raw file](https://github.com/cockroachdb/cockroach/blob/86f0469323c5a2cba0b08e1a6663a96d35250caf/pkg/server/serverpb/admin.proto#L766)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Comment appears to be stale in both the URL and in that it always runs on the leaseholder (for ranges not requiring lease, you run it on all replicas).
</blockquote></details>

Thanks for catching, fixed.

---

*[pkg/storage/store.go, line 4245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEXar4D_-9oEWUyaIDK-r1-4245:-LEzdITI3rRiu-qydEUp:b-xqxctv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/storage/store.go#L4245)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`consistency` is probably a little short, but we could change the log tags to add ` queue` to the `Name()`. Then again, that's probably not a good idea neither.
</blockquote></details>

I'm perfectly content with `consistencyChecker`, so let me know if you think changing it would be better. Otherwise I'll leave as is.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LF-01C3B8Dw6_n7nP5i:bbvktkx)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I'm happy with rangeid based addressing and the current naming of the
consistency queue. Can't wait to use this!

On Thu, Jun 14, 2018, 10:38 PM Alex Robinson <notifications@github.com>
wrote:

> Looks good! How about a simple unit test?
> Done.
>
> And add something to the debug overview page that gives the relevant curl
> invocations for an example or two.
>
> Done, but I think I've committed crimes against typescript in the process,
> so cc @couchand <https://github.com/couchand> to review that part.
> ------------------------------
>
> Review status: [image: :shipit:] complete! 0 of 0 LGTMs obtained
> ------------------------------
>
> *pkg/server/admin.go, line 1593 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEbUfvg5YD8PzE7l1CR:-LEzav3j6uCyjZJnuJqK:bld540p>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/server/admin.go#L1593>):*
> *Previously, bdarnell (Ben Darnell) wrote…*
>
> Maybe the request should contain a key instead of an ID, then. Or a
> specific node ID.
>
> Sorry for not responding to this yesterday. In my opinion that makes the
> interface much tougher to use. The only other status/admin server endpoint
> that takes keys is SpanStats, which I don't think is ever used directly
> by a human. When debugging, we (or at least I) usually work in terms of
> range IDs, at least partially due to the design of the existing debug
> endpoints. It would also add confusion about which types of keys are usable
> (how do they have to be encoded? pretty keys only?).
>
> I can switch it if you prefer, but I think after the recent changes to
> support running on non-leaseholder queues that the problem isn't as bad as
> it was before.
> ------------------------------
>
> *pkg/server/serverpb/admin.proto, line 553 at r2
> <https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEy-kSb54Qw-mukSEme:-LEzcHf62Bm3gcSh0b8F:b-j5yt41>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/86f0469323c5a2cba0b08e1a6663a96d35250caf/pkg/server/serverpb/admin.proto#L553>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> This is a bit more brute force than it needs to be (could also look up the
> range descriptor), but I like that this can work even if the meta ranges
> are wedged (and perhaps the queue is required to unwedge it). Consider
> putting that in a comment.
>
> As referenced the comment that Ben commented on in the original set of
> changes to admin.go, we can't look up the range descriptor from the range
> ID. We'd have to change the endpoint to take a key rather than a range ID
> in order to do a range descriptor lookup, which in my opinion makes the
> endpoint tougher to use. If I'm outvoted, though, I'm happy to change.
> ------------------------------
>
> *pkg/server/serverpb/admin.proto, line 557 at r2
> <https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEy-v4-Dempxl3JuUhh:-LEzdDji13qNN4x0AIkp:b-896fix>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/86f0469323c5a2cba0b08e1a6663a96d35250caf/pkg/server/serverpb/admin.proto#L557>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> Should explain what this is matched against.
>
> Done.
> ------------------------------
>
> *pkg/server/serverpb/admin.proto, line 558 at r2
> <https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEy-znNCRT3TK6_7BWq:-LEzcp3VBW43cZH4b4b5:byrhjpy>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/86f0469323c5a2cba0b08e1a6663a96d35250caf/pkg/server/serverpb/admin.proto#L558>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> Isn't this TODO done?
>
> Yes, thanks. Although I can't say I understand why it works now but it
> didn't work in my original version of this change.
> ------------------------------
>
> *pkg/server/serverpb/admin.proto, line 766 at r2
> <https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEy08psC_kTa45DX4c_:-LEzdQjPAOd6oZqjwu_o:b-jbu5lu>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/86f0469323c5a2cba0b08e1a6663a96d35250caf/pkg/server/serverpb/admin.proto#L766>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> Comment appears to be stale in both the URL and in that it always runs on
> the leaseholder (for ranges not requiring lease, you run it on all
> replicas).
>
> Thanks for catching, fixed.
> ------------------------------
>
> *pkg/storage/store.go, line 4245 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEXar4D_-9oEWUyaIDK-r1-4245:-LEzdITI3rRiu-qydEUp:b-xqxctv>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/storage/store.go#L4245>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> consistency is probably a little short, but we could change the log tags
> to add queue to the Name(). Then again, that's probably not a good idea
> neither.
>
> I'm perfectly content with consistencyChecker, so let me know if you
> think changing it would be better. Otherwise I'll leave as is.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LF-01C3B8Dw6_n7nP5i:bbvktkx>*
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/26554#issuecomment-397430487>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135NShD6LFYfK7jlxDypYFm-fKVl08ks5t8sm6gaJpZM4UhKIg>
> .
>
-- 

-- Tobias

bdarnell commented:
> I'd use that over the SQL version in all situations where we have UI access but not SQL access.

That's a weak justification - the reason we (sometimes) have UI access without SQL access today is that the UI is read-only and accessible anonymously. This operation isn't read-only, so once we have UI authentication it should be required here (and UI authentication will also imply some level of SQL access).

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/server/admin.go, line 1593 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEbUfvg5YD8PzE7l1CR:-LF33JC3Eqx9kp9KQYKh:bq4hpmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/server/admin.go#L1593)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Sorry for not responding to this yesterday. In my opinion that makes the interface much tougher to use. The only other status/admin server endpoint that takes keys is `SpanStats`, which I don't think is ever used directly by a human. When debugging, we (or at least I) usually work in terms of range IDs, at least partially due to the design of the existing debug endpoints. It would also add confusion about which types of keys are usable (how do they have to be encoded? pretty keys only?).

I can switch it if you prefer, but I think after the recent changes to support running on non-leaseholder queues that the problem isn't as bad as it was before.
</blockquote></details>

OK, range IDs are fine, then.

---

*[pkg/server/serverpb/admin.proto, line 758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEbVVgJ6RXomZvP89Nx:-LF33d7xEfgFjxEu8VPe:b-mhinie) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/server/serverpb/admin.proto#L758)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I thought the path pattern was more popular in REST API design, and there are a handful of other admin endpoints here that do something similar. I've changed it, though.
</blockquote></details>

It's popular, but IMHO misguided (it adds complexity for questionable aesthetic benefits in something that's not supposed to be visible anyway). Since we've started doing it in some places I'm fine with continuing the pattern for consistency, but I wouldn't generally encourage it.

---

*[pkg/ui/src/views/reports/containers/debug/index.tsx, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LF35lUR0M0vTLDnUXBv:-LF35lUR0M0vTLDnUXBw:b-i3e0mp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/ui/src/views/reports/containers/debug/index.tsx#L250)):*
> ```XML
>         <DebugTableRow title="Run Range Through Queue">
>           <tr className="debug-inner-table__row">
>             curl {window.location.protocol === 'https:' ? '-k' : ''} -H 'Content-Type: application/json' -X POST -d '{'{'}"queue": "consistencyChecker", "rangeId": 1}' {window.location.protocol + '//' + window.location.host + '/_admin/v1/queue'}
> ```

If this didn't use json it would be easy to make an HTML form for it here. (and we should really put this behind authentication so it's not going to be easy to do this with curl in the future)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LF34FMx-ks01UroXcqX:bnxghyv)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/ui/src/views/reports/containers/debug/index.tsx, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LF35lUR0M0vTLDnUXBv:-LFJxReuC-JpcjfnaqGA:bq77z3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/ui/src/views/reports/containers/debug/index.tsx#L250)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If this didn't use json it would be easy to make an HTML form for it here. (and we should really put this behind authentication so it's not going to be easy to do this with curl in the future)
</blockquote></details>

@couchand/@vilterp for thoughts on this. A form does seem like the ideal way to integrate this (rather than curl commands) given your point about auth.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Yeah, I guess it would make sense to make a little form, s.t. when you are logged in to the admin UI (currently requires starting cluster in secure mode, with `COCKROACH_EXPERIMENTAL_REQUIRE_WEB_LOGIN=true` env var), submitting the form would send the cookie along. I guess the form could be right in the debug index, or could have its own page… Thoughts @couchand?

To invoke it from the command line, someone could do it from the web first, then use [copy as curl](https://developers.google.com/web/updates/2015/05/replay-a-network-request-in-curl), which will include the cookie.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LFdvAqg4xLAM8XimuqD:bxz503u)*
<!-- Sent from Reviewable.io -->

couchand commented:
Definitely needs to be behind auth from my perspective.  Per the CLI/SQL/REST question: I'm not clear that this much of the plumbing should be tied to the HTTP endpoint, even if that is a (potentially) useful option.

---

Reviewed 3 of 8 files at r1, 8 of 10 files at r2, 3 of 3 files at r3.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/server/admin.go, line 1545 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxSaP4e_ZZlbrq8Sa:-LFrxSaP4e_ZZlbrq8Sb:b-i83xb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin.go#L1545)):*
> ```Go
> 	}
> 
> 	type nodeResponse struct {
> ```

see #26821

---

*[pkg/server/admin_test.go, line 1506 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxzog9AT6IMnxV88a:-LFrxzohCViheAFTMqYs:b-uwwcw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin_test.go#L1506)):*
> ```Go
> 			}
> 			if tc.expectedNonErrors != numNonErrors {
> 				t.Errorf("expected %d non-error details; got %d: %+v", tc.expectedNonErrors, numNonErrors, resp)
> ```

I think this assertion would fail a lot more obviously if the distinction between success cases with "error" responses and error cases were more explicit

---

*[pkg/server/serverpb/admin.proto, line 758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEbVVgJ6RXomZvP89Nx:-LFryjDO27-NqIXh1mBi:b-tabuzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/server/serverpb/admin.proto#L758)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's popular, but IMHO misguided (it adds complexity for questionable aesthetic benefits in something that's not supposed to be visible anyway). Since we've started doing it in some places I'm fine with continuing the pattern for consistency, but I wouldn't generally encourage it.
</blockquote></details>

This is a classic mismatch between REST and RPC...  I'm a big fan of putting entities in the URL, but it should be inverted from this.  It should be `/entity/child/operation` not `/operation/entity/child`.  But that's a big difference from the current structure, so probably not something to start on here...

---

*[pkg/server/serverpb/admin.proto, line 770 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrz6LXEernlU1DMpYf:-LFrz6LXEernlU1DMpYg:b-kzn8g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/serverpb/admin.proto#L770)):*
> ```protobuf
>   //   "rangeId": 10
>   // }
>   rpc Queue(QueueRequest) returns (QueueResponse) {
> ```

I'd like to bikeshed on the name.  Unlike all the other "RPC"s in this file, this one is actually a command, so the name should reflect that it's doing something commandy.

---

*[pkg/ui/src/views/reports/containers/debug/index.tsx, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LF35lUR0M0vTLDnUXBv:-LFs-MYV3PLMCjBQ795i:b-8z3dxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/ui/src/views/reports/containers/debug/index.tsx#L250)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

@couchand/@vilterp for thoughts on this. A form does seem like the ideal way to integrate this (rather than curl commands) given your point about auth.
</blockquote></details>

A form seems weird -- it's not like you're just waking up one morning and deciding which range to run a queue on.  Aren't you probably coming from another debug page?  Can we give you a button over there?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LFs-fsi8EKJ0T-m60Bg:b-2h9953)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/server/admin.go, line 1545 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxSaP4e_ZZlbrq8Sa:-LGAHqjS9PwX4feWtSze:b-hzasce) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin.go#L1545)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

see #26821
</blockquote></details>

yes that probably could use `(statusServer).iterateNodes()`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Sorry for the delayed response!

@couchand where would you prefer the plumbing to be? As stated earlier I'm personally a bit biased toward the debug UI since

1. It's where I have historically done a lot of my debugging
2. It's where using range IDs is most idiomatic. We rarely take range IDs as input elsewhere (especially in SQL), and I think it's important to do so for this endpoint because it requires the ability to pick a specific range from anywhere in the keyspace (including keys that can't be addressed via SQL).

But I'm also not very invested in the outcome, so if you have a strong preference of moving it I can abide (as long as targeting a specific range is still possible). If you don't, I'll keep it as an admin endpoint and put a form on a dedicated debug page linked from the `/#/debug` page.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/server/admin.go, line 1545 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxSaP4e_ZZlbrq8Sa:-LGRUb058IgFcmizXyk5:bmjau1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin.go#L1545)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

yes that probably could use `(statusServer).iterateNodes()`.
</blockquote></details>

Thanks, I'm sorry for missing that change! I'll clean that up pending what you think we should do overall.

---

*[pkg/server/admin_test.go, line 1506 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxzog9AT6IMnxV88a:-LGRW7Pr828-k00-cifN:bx4x394) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin_test.go#L1506)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think this assertion would fail a lot more obviously if the distinction between success cases with "error" responses and error cases were more explicit
</blockquote></details>

Do you think the API needs reworking?

---

*[pkg/server/serverpb/admin.proto, line 758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LEbVVgJ6RXomZvP89Nx:-LGRUzgc2tYF_kOqmkCY:b3ex4b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e44040e2b14c893333b31a9f7931c8e4b24513/pkg/server/serverpb/admin.proto#L758)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This is a classic mismatch between REST and RPC...  I'm a big fan of putting entities in the URL, but it should be inverted from this.  It should be `/entity/child/operation` not `/operation/entity/child`.  But that's a big difference from the current structure, so probably not something to start on here...
</blockquote></details>

Good point about the ordering here having been wrong. Agreed that this probably isn't the time to start changing, though. I'll stick with the short RPC-like endpoint.

---

*[pkg/server/serverpb/admin.proto, line 770 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrz6LXEernlU1DMpYf:-LGRWjmhAGqUxmeHEraL:b1r9kvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/serverpb/admin.proto#L770)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I'd like to bikeshed on the name.  Unlike all the other "RPC"s in this file, this one is actually a command, so the name should reflect that it's doing something commandy.
</blockquote></details>

Yeah, I'd been thinking the same thing but couldn't come up with anything I liked. I'd go with "RunQueue", but it has other meanings in software. 

How about "TraceQueue"? Or "QueueRange"? Let me know and I'll update it along with everything else based on your thoughts on the appropriate way of exposing this functionality.

---

*[pkg/ui/src/views/reports/containers/debug/index.tsx, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LF35lUR0M0vTLDnUXBv:-LGRVB_g7q3AfK1ofK9i:b-z1qpqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/ui/src/views/reports/containers/debug/index.tsx#L250)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

A form seems weird -- it's not like you're just waking up one morning and deciding which range to run a queue on.  Aren't you probably coming from another debug page?  Can we give you a button over there?
</blockquote></details>

I think I'd most likely be coming from the logs, from the range status debug page, or from the Queues dashboard (in the case that a given queue is backed up and moving very slowly on all ranges, not just one).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LGRTaI-4y95QLBXvZzs:b-ks598o)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
ping @couchand

couchand reviewed (changes requested):
And now I'm the one apologizing for the delayed response @a-robinson!

I may have to wait to see how this shakes out after moving over to using the `iterateNodes` helper.  That alone might clear up my concern.

Basically I'm just worried that we keep throwing code in `pkg/server` and nobody really owns it and a lot of it doesn't really belong there (I'll admit to being just as guilty of this as the next person).  For instance, with this change, the API endpoint should of course be in `pkg/server`, but the actual logic to find a queue and enqueue a range should probably be somewhere else, right?

I'm not sure how the range ID question fits into this.  It sounds like it makes sense to use that as the unique identifier here.  I'm curious why it is it's not used elsewhere, and was that an intentional choice?  In any case that seems orthogonal to the question of how to factor and expose the functionality.

> Reviewed 10 of 11 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LHnIk6e8hfrqOuRj9rX:b-f4ea9f)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/admin_test.go, line 1506 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxzog9AT6IMnxV88a:-LHn0V8_CXTo6ohSGVTJ:b-vp8ndv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin_test.go#L1506)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Do you think the API needs reworking?
</blockquote></details>

No it's just the test case structure: making it more explicit that the three cases are "replica", "leaseholder", and "nada", rather than the somewhat ambiguous constants 3, 1, and 0 for `expectedNonErrors`.
___
*[pkg/server/serverpb/admin.proto, line 770 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrz6LXEernlU1DMpYf:-LHn1LbR5CUOn7Ocfjlz:btustu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/serverpb/admin.proto#L770)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Yeah, I'd been thinking the same thing but couldn't come up with anything I liked. I'd go with "RunQueue", but it has other meanings in software. 

How about "TraceQueue"? Or "QueueRange"? Let me know and I'll update it along with everything else based on your thoughts on the appropriate way of exposing this functionality.
</blockquote></details>

proposal: `EnqueueRange`
___
*[pkg/ui/src/views/reports/containers/debug/index.tsx, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LF35lUR0M0vTLDnUXBv:-LHn2Xwr1xnK0Y_ZrYf6:by68k9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/ui/src/views/reports/containers/debug/index.tsx#L250)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I think I'd most likely be coming from the logs, from the range status debug page, or from the Queues dashboard (in the case that a given queue is backed up and moving very slowly on all ranges, not just one).
</blockquote></details>

I think some way to do this in-context would be ideal, but if it would be hard for us to specify exactly what we want and where than maybe just a plain form will work for now.


<!-- Sent from Reviewable.io -->

a-robinson reviewed (commented):
> Basically I'm just worried that we keep throwing code in pkg/server and nobody really owns it and a lot of it doesn't really belong there (I'll admit to being just as guilty of this as the next person). For instance, with this change, the API endpoint should of course be in pkg/server, but the actual logic to find a queue and enqueue a range should probably be somewhere else, right?

Maybe? But right now the only endpoints that try to do anything with a provided range ID are other status/admin server endpoints, so it fits there pretty well from a code grouping perspective.

> I'm not sure how the range ID question fits into this. It sounds like it makes sense to use that as the unique identifier here. I'm curious why it is it's not used elsewhere, and was that an intentional choice?

I believe the logic has been that SQL interfaces should use database/table/index names and primary keys or select statements for accessing data, so those are preferred over range IDs there. For debugging-related endpoints outside of SQL, range IDs have always been preferred AFAIK. The conflict arises when we start exposing more of this stuff via SQL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LICSyFKC_wIQ5AT9F01:bka9sgh)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/admin.go, line 1545 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxSaP4e_ZZlbrq8Sa:-LICMvA-83_79kooCLTc:b2ybp8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin.go#L1545)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Thanks, I'm sorry for missing that change! I'll clean that up pending what you think we should do overall.
</blockquote></details>

Done, although it didn't fit quite as cleanly as the other uses due to being part of the adminServer instead of the statusServer.
___
*[pkg/server/admin_test.go, line 1506 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxzog9AT6IMnxV88a:-LICOEMmEXVduZZCV74r:b-yw8ca2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin_test.go#L1506)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

No it's just the test case structure: making it more explicit that the three cases are "replica", "leaseholder", and "nada", rather than the somewhat ambiguous constants 3, 1, and 0 for `expectedNonErrors`.
</blockquote></details>

Any better?
___
*[pkg/server/serverpb/admin.proto, line 770 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrz6LXEernlU1DMpYf:-LI8Cn0RB9QAcMifhCMw:b-6008xe) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/serverpb/admin.proto#L770)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

proposal: `EnqueueRange`
</blockquote></details>

Beautiful. Done.
___
*[pkg/ui/src/views/reports/containers/debug/index.tsx, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LF35lUR0M0vTLDnUXBv:-LICSVhcAI7drMSXAq0Q:bd0e763) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/ui/src/views/reports/containers/debug/index.tsx#L250)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think some way to do this in-context would be ideal, but if it would be hard for us to specify exactly what we want and where than maybe just a plain form will work for now.
</blockquote></details>

If I just write a plain HTML form will that plug into whatever authentication we have? I'm at a bit of a loss as to how to properly do this in our typescript/react/redux setup, so if you can point to anything that I can copy from I'd be very appreciative.


<!-- Sent from Reviewable.io -->

couchand reviewed (dismissed):
LGTM, pending a decision on the UI side.

I think there are one or two responsibilities that can be split out of the server package (the iterate nodes stuff, for instance, and perhaps the iterate stores stuff), but it's probably not something to address today.

> Reviewed 13 of 14 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-:-LINLXSs0-6BVhnBMLbv:b-khw9z1)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/admin.go, line 1545 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxSaP4e_ZZlbrq8Sa:-LINGhHG2UwHp8Qh7hUn:b-3ep6zp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin.go#L1545)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Done, although it didn't fit quite as cleanly as the other uses due to being part of the adminServer instead of the statusServer.
</blockquote></details>

oh fun.

cf. #24989
___
*[pkg/server/admin_test.go, line 1506 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LFrxzog9AT6IMnxV88a:-LINH48ME0h53FrwISrs:bv0kb09) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/server/admin_test.go#L1506)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Any better?
</blockquote></details>

Oh yeah that's great, thanks!
___
*[pkg/ui/src/views/reports/containers/debug/index.tsx, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LF35lUR0M0vTLDnUXBv:-LINI4kH-_9TQ1ndHsGe:b-hp8btm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/ui/src/views/reports/containers/debug/index.tsx#L250)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

If I just write a plain HTML form will that plug into whatever authentication we have? I'm at a bit of a loss as to how to properly do this in our typescript/react/redux setup, so if you can point to anything that I can copy from I'd be very appreciative.
</blockquote></details>

Since it's sending to the same host the browser will add the cookie header implicitly, there's nothing in particular you need to do there.

We accept JSON and protobufs, but it doesn't look like we accept the HTML default form encoding, so you will need to write a little glue code to send off the request.  Take a look at, for instance, `src/util/api` for the type of code you'll write to make the actual request.  The form itself you'll put in a React component, wiring the form submit event into the RPC call.  Then you can drop that component on some relevant new or existing page.

If you need more specific details here I'm happy to chat further.


<!-- Sent from Reviewable.io -->

a-robinson commented:
I have other more important work to get in before code freeze than the debug UI page for this, but I want to make sure that the store/server code for this is in. I'm going to merge this without the UI commit for now.

a-robinson commented:
bors r+

a-robinson commented:
bors r+

a-robinson commented:
bors r+

a-robinson commented:
Opened https://github.com/cockroachdb/cockroach/issues/28538 to track the debug page.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26554)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ui/src/views/reports/containers/debug/index.tsx, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26554#-LF35lUR0M0vTLDnUXBv:-LLGGinYAqZtLP3fwKVN:b5aevye) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f0d981187bd7c0abf42e82d05c1c78005536011/pkg/ui/src/views/reports/containers/debug/index.tsx#L250)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Since it's sending to the same host the browser will add the cookie header implicitly, there's nothing in particular you need to do there.

We accept JSON and protobufs, but it doesn't look like we accept the HTML default form encoding, so you will need to write a little glue code to send off the request.  Take a look at, for instance, `src/util/api` for the type of code you'll write to make the actual request.  The form itself you'll put in a React component, wiring the form submit event into the RPC call.  Then you can drop that component on some relevant new or existing page.

If you need more specific details here I'm happy to chat further.
</blockquote></details>

Was it intentional that these lines got removed? Cause I was trolling about how I'm supposed to be using this new page.


<!-- Sent from Reviewable.io -->

a-robinson commented:
> Was it intentional that these lines got removed? Cause I was trolling about how I'm supposed to be using this new page.

Yeah, it was because they wouldn't work on any clusters with login enabled. It looks like you already found https://github.com/cockroachdb/cockroach/issues/28538, which is tracking a functional form for this purpose for 2.1. If Pete/Andrew/Matt find any time it'd probably take them 10% as long to do it as it'd take me to figure out how to set up such a form, but since it's just adding a debug page it's very low risk and can be done late in the cycle if needed.
